File: python/trees/0105-construct-binary-tree-from-preorder-and-inorder-traversal.py
from typing import List

from structures import TreeNode


def buildTree(self, preorder: List[int], inorder: List[int]) -> TreeNode:
    if len(preorder) == 0 or len(inorder) == 0: return None
    
    val = preorder.pop(0)
    index_val = inorder.index(val)
    
    
    left, right = None, None
    
    if index_val > 0:
        left = self.buildTree(preorder, inorder[:index_val])
        
    if index_val < len(inorder):
        right = self.buildTree(preorder, inorder[index_val + 1:])
        
    return TreeNode(val, left, right)

File: cpp/trees/0105-construct-binary-tree-from-preorder-and-inorder-traversal.cpp
#include <algorithm>
#include <vector>
#include <structures.h>

using namespace std;


TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
    return build(preorder, inorder, 0, inorder.size() - 1);
}

TreeNode* build(vector<int>& preorder, vector<int>& inorder, int start, int end) {
    if (start > end) return nullptr;
    
    int val = preorder[0];
    preorder.erase(preorder.begin());
    
    TreeNode* root = new TreeNode(val);
    
    if (start != end) {
        int index = find(inorder.begin() + start, inorder.begin() + end, val) - inorder.begin();
        if (index > start) root->left = build(preorder, inorder, start, index - 1);
        if (index < end) root->right = build(preorder, inorder, index + 1, end);
    }
    
    return root;
}